libs/payments/authenticated_payment_state.go

```diff
@@ -0,0 +1,80 @@
+package payments
```

Comment:
We intentionally moved them for 2 reasons:

1. So that the conversion from `PaymentState` to `AuthenticatedPaymentState` can be strictly controlled, preventing interactions with unverified data.
2. So that the request and response types can be shared among the 3 systems involved here.

Reply:
Maybe we can discuss this once we review the nitro-payments-dev branch as they seem like strong reasons not have a shared types.

Reply:
> 2\. So that the request and response types can be shared among the 3 systems involved here.

This decision is limiting in the mid- and long-term. It blurs the external boundaries between the 3 systems, making it hard to change them (e.g. gradually update the output of one without breaking the input of another).

I recommend reconsidering this decision before the code gets accepted.

Reply:
normally, this would be good practice.

however in this case there is actually only one system, treating them as separate only invites the opportunity for subtle discrepancies in a use case that allows no margin for such errors. this is internal tooling and we can easily update all components of the system in lockstep. the only external boundary here is the input payout report - everything else is internal. I'd actually go as far to say we should introduce the requirement that the same revision of code is used between the CLI and enclave.

Reply:
Normally, I would be in favor of the flexibility you're describing. However, because this component will distribute money I think we should intentionally preclude the possibility of interface proliferation between the tooling, workers, and enclave executor. Tight coupling in this case, while restricting how we make changes, eliminates an especially dangerous category of error. That is, payment bugs that can result from handling diverse interfaces. As much as possible, the "separate" pieces of this system should be integrated and the consequences of changes in any piece should be considered throughout the system.

I do agree that we should probably move `lib/payments` into `service/payments` to avoid the appearance of interoperability with other systems.

Reply:
While I do see what you both @evq and @Sneagan are saying, I respectfully disagree.

The argument that this is a single system but running as three processes is weak. It can only be viewed as a single system, if changes to all the three components are applied exactly at the same time. That is, you'd have to fully stop all instances of all the systems, and only then bring up a new version for all of them. And while I can image that being possible, that further complicates other scenarios such as roll backs. While it might be still possible to do in manual mode, and hence make sure all instances are down before replacements are getting up, it certainly will complicate automation.

The argument that this artificial tight coupling can reduce the likelihood of software defects is also weak. Transport layer is not usually the place where issues originate. Mistakes more often occur in business logic and inconsistent data. Good design, careful implementation, good tests and thorough testing are the means to reduce bugs. And tight coupling is among the very things that complicate the means through which the goal is achieved.

Reply:
It seems difficult to reason about the CLI, Worker and Payment Service as one system from both a technical and domain perspective some of which have been mentioned in other comments.

In short, each piece has very specific responsibilities and provides an explicit interface to interact with and sharing code between them could be very limiting. For example, these seem are very specific to the internal workings of the payment service and interaction should only happen through the payment service interface -
* [payment_state.go](https://github.com/brave-intl/bat-go/pull/2001/files#diff-ac039de62688ba856cac8e1ed82199b3dd4e1018aebd1812665a5e357b5b395f) 
* [qldb_types.go](https://github.com/brave-intl/bat-go/pull/2001/files#diff-0e8d5c73d011a95a48905afe1796c2785691d4ea74911c3c83cb3de2a92051e4) 
* [transitions.go](https://github.com/brave-intl/bat-go/pull/2001/files#diff-79141be492c445d0fa6f0700038c98f9c88821868a6cc2d4246d0213795c3534R9) 

If we add these models to a shared lib we leek out  the internal details and reduce cohesion. This also seems to go against @snegan point about tight control.

Both [submit.go](https://github.com/brave-intl/bat-go/pull/2001/files#diff-3a515cc5d1ad655d2da14de1837d0ebe09c949eade6c2db1e7c1896537d01a60) and [prepare.go](https://github.com/brave-intl/bat-go/pull/2001/files#diff-bab2bfb1a3b2b04c7ac425e04db7219735d192a1f69f2b9bbb6d4634fef9c7f8) contain request and response types which look like simple dto's, given their simplicity it seems like unnecessary coupling and potential for issues if we want to update the API.

[interfaces.go](https://github.com/brave-intl/bat-go/pull/2001/files#diff-32e1d49edcadd5c2f2746f0b18e57bc48026233d2e49f1f2f25fe7dade33ab6fR8) Its not clear why this is being shared and if it is even needed.

Details could be a candidate for a shared model but at this point, are we sure what models we want to share? It includes validation tags which makes assumptions about the use of a specific lib i.e. `govalidator` which is not great practice in library code.

The fact this is _internal_ seems to suggest a shared library which is imported into other systems may not be the best place for this code.

If the intention is for them to be seen and treated as one system would the cli, workers and payment service not reside under a single package/module?

Reply:
@clD11 to the contrary, reasoning about them as one system is essential. only in concert do they deliver the required security properties.

I am in favor of consolidating the cli, workers and payment service into a single module. the vast majority of logic lives in the payment service - the workers and cli are fairly small and exist only to bootstrap the enclave and sling requests from a report to the enclave. like I said before, the only real external interface here is the payout report. treating anything else ( payment API, redis streams ) as an external interface would be quite a mistake at this stage! 

re: updating the API - common structures does not preclude us from versioning the API.

@pavelbrm I appreciate the concern but I need to reiterate that this is a bit of an unusual case in that it would be no problem at all to bring down all instances of everything in order to coordinate changes. a new revision of the code literally cannot begin to operate without first having multiple operators manually collaborate to provide the unseal shares. that will not change, it's fundamental to the security model involved here...

anyway, this is just the first phase of refactoring - don't get too worried about the exact shape of things as they stand here as we are continuing to iterate.